toolchain/parse/context.h

```diff
@@ -44,6 +44,14 @@ class Context {
     NamedConstraint,
   };
 
+  // Used for restricting ordering of `package` and `import` directives.
+  enum class PackagingState : int8_t {
+    PackageOrImportsAllowed,
+    PackageEncountered,
+    ImportsAllowed,
+    PackageOrImportsInvalid,
+  };
```

Comment:
I find the names of these enumerators confusing, because they're describing different kinds of information (the first and third enumerators say that something is allowed, the fourth says that something is disallowed, and the second describes our position in parsing). I'd find this easier to follow if these were all phrased as the same kind of information. For example:
```suggestion
  enum class PackagingState : int8_t {
    StartOfFile,
    AfterPackageDeclaration,
    AfterImportDeclaration,
    AfterNonPackagingDeclaration,
  };
```

It also seems to me that the second and third states aren't really different -- in either state, we can have imports but not a package declaration. I note that you currently use this distinction to produce a custom diagnostic for:

```carbon
package A;
package B;
```

that says we can't have two package declarations, whereas you get a different diagnostic (package declaration must be at the start of the file) for:

```carbon
import A;
package B;
```
and for
```carbon
package A;
import B;
package C;
```

... but I think it would be reasonable, and more consistent, to produce the "package declaration must be at the start of the file" diagnostic for all of these cases rather than special-casing two package declarations in a row but not two package declarations in general.

Alternatively, we could stay in the "after package declaration" state when we hit an import after a package, so that `package...; import...; package...;` also produces the "multiple package declarations" error instead of the "package declaration must be at the start of the file" error, but we'd still get the "package declaration must be at the start of the file" error for `package ...; fn ...; package ...;` in that case, so it'd still be a bit inconsistent. Maybe we could track three independent bits -- whether we've seen any package declarations, whether we've seen any import declarations, and whether we've seen any other kinds of declarations.

Reply:
My concern was that people may be confused by:

```
package Foo;
package Foo;
```

resulting in a "must be at the start of the file". And I could make it longer ("there must be only one, and it must be at the start of the file") but that felt a little too much. So that's how I ended up with the current setup, even though the `package import package` is a similar issue (but, clearer that the second `package` isn't at the top)

Anyways, doing the name changes.

Reply:
Or rather, maybe I should clarify.

First, we refer to `package` and `import` as "directives" instead of "declarations", so I'm trying to be careful about phrasing there.

Then we have:

```
// Foo
package Foo;
```

This isn't technically at the start of the file, but it is before declarations. So I feel I shouldn't just say "must be at the start of the file".

I could maybe say first non-comment line; but can we commit to never adding anything that would be valid above? And even then, is it as clear on a duplicate entry that's still adjacent to the first non-comment line?

Reply:
I'll note, kind of tempting is a "first processed item is here" kind of diagnostic, though I wasn't sure it's worth the additional state.

Although, maybe it's worth it...

Reply:
I think pointing at the previous line helps to clarify what we're talking about; I like the new approach of tracking where we switched state.